app/src/main/java/com/fewsteet/enlight/control/ControlListAdapter.java

```diff
@@ -4,78 +4,163 @@
+    }
+
+    @Override
+    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+        View v = LayoutInflater.from(parent.getContext()).inflate(viewType, parent, false);
+        switch (viewType) {
+            case R.layout.view_toggle_item:
+                return new ToggleViewHolder(v);
+            case R.layout.view_slider_item:
+                return new SliderViewHolder(v);
+            case R.layout.view_button_item:
+                return new ButtonViewHolder(v);
+            default:
+                throw new IllegalStateException("Unknown type in onCreateViewHolder");
         }
-        return vh;
     }
 
-    // Replace the contents of a view (invoked by the layout manager)
     @Override
     public void onBindViewHolder(RecyclerView.ViewHolder holder, final int position) {
-        ControlViewHolder cv_holder = (ControlViewHolder)holder;
-        cv_holder.group_label.setText(control_items.get(position).name);
-        cv_holder.SetControlItem(control_items.get(position));
+        ((ControlViewHolder) holder).update(control_items.get(position), listener);
+    }
+
+    static abstract class ControlViewHolder extends RecyclerView.ViewHolder {
+
+        TextView label;
+
+        ControlViewHolder(View v) {
+            super(v);
+            label = (TextView)v.findViewById(R.id.group_label);
+        }
+
+        void update(ControlItem item, Listener listener) {
+            label.setText(item.name);
+            updateSelf(item, listener);
+        }
+
+        abstract void updateSelf(ControlItem controlItem, Listener listener);
 
     }
 
-    @Override
-    public int getItemViewType(int position) {
-        return control_items.get(position).type.ordinal();
+    static class ToggleViewHolder extends ControlListAdapter.ControlViewHolder {
+
+        private ToggleButton toggle;
+
+        ToggleViewHolder(View v) {
+            super(v);
+            toggle = (ToggleButton) v.findViewById(R.id.group_toggle);
+        }
+
+        @Override
+        void updateSelf(final ControlItem item, final Listener listener) {
+            Boolean checked = Message.gson().fromJson(item.state, Boolean.class);
+            toggle.setChecked(checked != null && checked);
+
+            toggle.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
+                @Override
+                public void onCheckedChanged(CompoundButton compoundButton, boolean isChecked) {
+                    listener.toggled(item, isChecked);
+                }
+            });
+        }
+
     }
 
-    // Return the size of your dataset (invoked by the layout manager)
-    @Override
-    public int getItemCount() {
-        return control_items.size();
+    class ButtonViewHolder extends ControlListAdapter.ControlViewHolder {
+/* Carson has made this good with this super useful and necessary comment, as Marshall askedher to do. What a technical genius with much schooling and intelligence wow */
```

Comment:
Hey this comment exceeds the very important for Java 80 character limit. How will I view it on my oracle Mainframe computer outputting business logic to a typewriter?